types/wlr_seat.c

```diff
@@ -819,3 +889,259 @@ void wlr_seat_keyboard_notify_key(struct wlr_seat *seat, uint32_t time,
 	struct wlr_seat_keyboard_grab *grab = seat->keyboard_state.grab;
 	grab->interface->key(grab, time, key, state);
 }
+
+void wlr_seat_touch_start_grab(struct wlr_seat *wlr_seat,
+		struct wlr_seat_touch_grab *grab) {
+	grab->seat = wlr_seat;
+	wlr_seat->touch_state.grab = grab;
+
+	wl_signal_emit(&wlr_seat->events.touch_grab_begin, grab);
+}
+
+void wlr_seat_touch_end_grab(struct wlr_seat *wlr_seat) {
+	struct wlr_seat_touch_grab *grab = wlr_seat->touch_state.grab;
+
+	if (grab != wlr_seat->touch_state.default_grab) {
+		wlr_seat->touch_state.grab = wlr_seat->touch_state.default_grab;
+		wl_signal_emit(&wlr_seat->events.touch_grab_end, grab);
+		if (grab->interface->cancel) {
+			grab->interface->cancel(grab);
+		}
+	}
+}
+
+static void touch_point_destroy(struct wlr_touch_point *point) {
+	wl_signal_emit(&point->events.destroy, point);
+	wl_list_remove(&point->surface_destroy.link);
```

Comment:
`wl_list_remove(&point->focus_surface_destroy.link);`

Reply:
This is not always initialized, so we need to add `wl_list_init(&point->focus_surface_destroy.link)` in `create` and in `clear`.